modules/core/m_mode.c

```diff
@@ -368,53 +377,83 @@ ms_bmask(struct MsgBuf *msgbuf_p, struct Client *client_p, struct Client *source
 						parv[3][0], s, forward);
 		}
 
-		if(add_id(fakesource_p, chptr, s, forward, banlist, mode_type))
+		mask = s;
+		if (extended) {
+			when = atol(strtok(NULL, " "));
+			who = strtok(NULL, " ");
+			if (who == NULL)
+			{
+				/* EBMASK params don't divide by 3, so we have an incomplete chunk */
+				sendto_realops_snomask(SNO_GENERAL, L_NETWIDE,
+					"Link %s dropped, invalid EBMASK chunk", source_p->name);
+				snprintf(squitreason, sizeof squitreason, "Invalid EBMASK chunk");
+				exit_client(client_p, client_p, client_p, squitreason);
+				return;
+			}
+
+			arglen = sprintf(degrade_ptr, "%s ", mask);
+			degrade_ptr += arglen;
+		}
+
+		if((banptr = add_id(fakesource_p, chptr, mask, forward, banlist, mode_type)) != NULL)
 		{
+			banptr->when = when;
+			rb_free(banptr->who);
+			banptr->who = rb_strdup(who);
+
 			/* this new one wont fit.. */
 			if(mlen + MAXMODEPARAMS + plen + tlen > BUFSIZE - 5 ||
 			   modecount >= MAXMODEPARAMS)
 			{
-				*mbuf = '\0';
-				*(pbuf - 1) = '\0';
-				sendto_channel_local(fakesource_p, mems, chptr, "%s %s", modebuf, parabuf);
+				*output_ptr = '\0';
+				*(param_ptr - 1) = '\0';
+				sendto_channel_local(fakesource_p, mems, chptr, "%s %s", output, parabuf);
 
-				mbuf = modebuf + mlen;
-				pbuf = parabuf;
+				output_ptr = output + mlen;
+				param_ptr = parabuf;
 				plen = modecount = 0;
 			}
 
 			if (forward != NULL)
 				forward[-1] = '$';
 
-			*mbuf++ = parv[3][0];
-			arglen = sprintf(pbuf, "%s ", s);
-			pbuf += arglen;
+			*output_ptr++ = parv[3][0];
+			arglen = sprintf(param_ptr, "%s ", mask);
+			param_ptr += arglen;
 			plen += arglen;
 			modecount++;
 		}
 
-	      nextban:
-		s = t;
-
-		if(s != NULL)
-		{
-			if((t = strchr(s, ' ')) != NULL)
-			{
-				*t++ = '\0';
-
-				while(*t == ' ')
-					t++;
-			}
-		}
+		s = strtok(NULL, " ");
 	}
 
 	if(modecount)
 	{
-		*mbuf = '\0';
-		*(pbuf - 1) = '\0';
-		sendto_channel_local(fakesource_p, mems, chptr, "%s %s", modebuf, parabuf);
+		*output_ptr = '\0';
+		*(param_ptr - 1) = '\0';
+		*(degrade_ptr - 1) = '\0';
+		sendto_channel_local(fakesource_p, mems, chptr, "%s %s", output, parabuf);
+	}
+
+	if (extended) {
+		sendto_server(client_p, chptr, CAP_EBMASK | CAP_TS6 | needcap, NOCAPS, ":%s EBMASK %ld %s %s :%s",
+			      source_p->id, (long) chptr->channelts, chptr->chname, parv[3], parv[4]);
+		sendto_server(client_p, chptr, CAP_TS6 | needcap, CAP_EBMASK, ":%s BMASK %ld %s %s :%s",
+			      source_p->id, (long) chptr->channelts, chptr->chname, parv[3], degrade);
```

Comment:
```suggestion
		*output_ptr = '\0';
		*(param_ptr - 1) = '\0';
		sendto_channel_local(fakesource_p, mems, chptr, "%s %s", output, parabuf);
	}

	if (extended) {
		*(degrade_ptr - 1) = '\0';

		sendto_server(client_p, chptr, CAP_EBMASK | CAP_TS6 | needcap, NOCAPS, ":%s EBMASK %ld %s %s :%s",
			      source_p->id, (long) chptr->channelts, chptr->chname, parv[3], parv[4]);
		sendto_server(client_p, chptr, CAP_TS6 | needcap, CAP_EBMASK, ":%s BMASK %ld %s %s :%s",
			      source_p->id, (long) chptr->channelts, chptr->chname, parv[3], degrade);
```

Move this to the right spot; where it is now will result in an out of bounds write in the `extended == false` case.